File: integ_with_arrow_sam/plasma_access.cc
#include "plasma_access.h"

arrow::Status PutRecordBatchToPlasma(
    const std::shared_ptr<arrow::RecordBatch>& record_batch,
    plasma::ObjectID object_id) {
    // Get size of the Record Batch
    arrow::io::MockOutputStream mock_output_stream;
    std::shared_ptr<arrow::ipc::RecordBatchWriter> record_batch_writer;
    arrow::Status status;
    status = arrow::ipc::RecordBatchStreamWriter::Open(
        &mock_output_stream, record_batch->schema(), &record_batch_writer);
    if (!status.ok()) {
        std::cerr
            << "arrow::ipc::RecordBatchStreamWriter::Open failed with error: "
            << status.ToString() << std::endl;
    }
    status = record_batch_writer->WriteRecordBatch(*record_batch);
    if (!status.ok()) {
        std::cerr << "record_batch_writer->WriteRecordBatch failed with error: "
                  << status.ToString() << std::endl;
    }
    int64_t data_size = mock_output_stream.GetExtentBytesWritten();

    // Start up a client and connect it to Plasma
    plasma::PlasmaClient client;
    ARROW_CHECK_OK(client.Connect("/tmp/plasma"));

    // Create an object with the given Object ID
    std::shared_ptr<Buffer> buf;
    ARROW_CHECK_OK(client.Create(object_id, data_size, nullptr, 0, &buf));

    // Write the Record Batch into the object
    arrow::io::FixedSizeBufferWriter buffer_writer(buf);
    status = arrow::ipc::RecordBatchStreamWriter::Open(
        &buffer_writer, record_batch->schema(), &record_batch_writer);
    if (!status.ok()) {
        std::cerr
            << "arrow::ipc::RecordBatchStreamWriter::Open failed with error: "
            << status.ToString() << std::endl;
    }
    status = record_batch_writer->WriteRecordBatch(*record_batch);
    if (!status.ok()) {
        std::cerr << "record_batch_writer->WriteRecordBatch failed with error: "
                  << status.ToString() << std::endl;
    }

    // Seal the object
    ARROW_CHECK_OK(client.Seal(object_id));

    // Disconnect the client
    ARROW_CHECK_OK(client.Disconnect());

    return status;
}

std::shared_ptr<arrow::RecordBatch> GetRecordBatchFromPlasma(
    plasma::ObjectID object_id, plasma::PlasmaClient& client) {
    plasma::ObjectBuffer object_buffer;
    ARROW_CHECK_OK(client.Get(&object_id, 1, -1, &object_buffer));

    auto buffer = object_buffer.data;

    arrow::io::BufferReader buffer_reader(buffer);
    std::shared_ptr<arrow::ipc::RecordBatchReader> record_batch_stream_reader;
    ARROW_CHECK_OK(arrow::ipc::RecordBatchStreamReader::Open(
        &buffer_reader, &record_batch_stream_reader));

    std::shared_ptr<arrow::RecordBatch> record_batch;
    arrow::Status status = record_batch_stream_reader->ReadNext(&record_batch);

    return record_batch;
}

File: integ_with_arrow_sam/plasma_access.py
import pyarrow as pa

def put_df_to_plasma(client, df, object_id):
    """
    Precondition: the Plasma Object Store has been opened.
    e.g. by: plasma_store -m 1000000000 -s /tmp/plasma
    """
    record_batch = pa.RecordBatch.from_pandas(df)
    # Get size of record batch and schema
    mock_output_stream = pa.MockOutputStream()
    stream_writer = pa.RecordBatchStreamWriter(mock_output_stream,
                                               record_batch.schema)
    stream_writer.write_batch(record_batch)
    data_size = mock_output_stream.size()

    # Allocate a buffer in the object store for the serialized DataFrame
    buf = client.create(object_id, data_size)

    # Write the serialized DataFrame to the object store
    stream = pa.FixedSizeBufferWriter(buf)
    stream_writer = pa.RecordBatchStreamWriter(stream, record_batch.schema)
    stream_writer.write_batch(record_batch)

    # Seal the object
    client.seal(object_id)


def get_record_batch_from_plasma(object_id, client):
    """ https://arrow.apache.org/docs/python/plasma.html#getting-pandas-dataframes-from-plasma
    """
    [buffer] = client.get_buffers([object_id])
    buffer_reader = pa.BufferReader(buffer)
    reader = pa.RecordBatchStreamReader(buffer_reader)
    record_batch = reader.read_next_batch()

    return record_batch
